File: packages/slam/src/Slam.cpp
/*
 * Copyright (C) 2009 Robotics at Maryland
 * Copyright (C) 2009 Jonathan Sternberg <jsternbe@umd.edu>
 * All rights reserved.
 *
 * Author: Jonathan Sternberg <jsternbe@umd.edu>
 * File:  packages/slam/src/Slam.cpp
 */

// STD Libraries
#include <cassert>

// Library Includes
#include <boost/foreach.hpp>
#include <boost/bind.hpp>

// Project Includes
#include "core/include/EventHub.h"
#include "core/include/SubsystemMaker.h"
#include "core/include/EventConnection.h"
#include "vision/include/Events.h"

#include "slam/include/Slam.h"

// Register slam into the maker subsystem
//RAM_CORE_REGISTER_SUBSYSTEM_MAKER(ram::slam::Slam, Slam);

namespace ram {
namespace slam {

Slam::Slam(core::ConfigNode config, core::SubsystemList deps) :
    ISlam(config["name"].asString(),
	  core::Subsystem::getSubsystemOfType<core::EventHub>(deps)),
    m_controller(core::Subsystem::getSubsystemOfType<
		 control::IController>(deps)),
    m_vehicle(core::Subsystem::getSubsystemOfType<
	      vehicle::IVehicle>(deps))
{
    // Find all of the objects
    core::ConfigNode objectConfig = config["objects"];

    // They should be in a map with the object name and the default vector2
    core::NodeNameList objectList = config["objects"].subNodes();

    // Iterate over all of the objects and add them to the map
    core::NodeNameListIter beg = objectList.begin();
    core::NodeNameListIter end = objectList.end();
    
    for ( ; beg != end; ++beg) {
	double x = objectConfig[*beg][0].asDouble();
	double y = objectConfig[*beg][1].asDouble();
	m_objects[*beg] = math::Vector2(x, y);
    }

    // Subscribe to all of the vision events
    core::EventConnectionPtr connection = subscribe(
	vision::EventType::LIGHT_FOUND,
        boost::bind(&Slam::onLightUpdate, this, _1));
    m_connections.push_back(connection);
}

Slam::~Slam()
{
    BOOST_FOREACH(core::EventConnectionPtr event, m_connections) {
	event->disconnect();
    }
}

math::Vector2 Slam::getObjectPosition(std::string name)
{
    ObjectMapIter iter = m_objects.find(name);

    // Assert that this object must be found
    assert(iter != m_objects.end());
    return iter->second;
}

math::Vector2 Slam::getRelativePosition(std::string name)
{
    ObjectMapIter iter = m_objects.find(name);

    // Assert that this object must be found
    assert(iter != m_objects.end());
    return m_vehicle->getPosition() - iter->second;
}

bool Slam::hasObject(std::string name)
{
    // count() will return 1 if it exists, 0 if it doesn't
    return (m_objects.count(name) == 1) ? true : false;
}

void Slam::onLightUpdate(core::EventPtr event)
{
}

void Slam::onPipeUpdate(core::EventPtr event)
{
}

void Slam::onBwireUpdate(core::EventPtr event)
{
}

void Slam::onBinUpdate(core::EventPtr event)
{
}

void Slam::onTargetUpdate(core::EventPtr event)
{
}

} // namespace slam
} // namespace ram

File: tools/simulator/src/sim/slam.py
# Copyright (C) 2009 Maryland Robotics Club
# Copyright (C) 2009 Jonathan Sternberg <jsternbe@umd.edu>
# All rights reserved.
#
# Author: Jonathan Sternberg <jsternbe@umd.edu>
# File:   tools/simulator/src/sim/slam.py

# Library Imports
import ogre.renderer.OGRE as ogre

# Project Imports
import ext.core as core
import ext.control as control
import ext.math as math
import ext.slam as slam
import ext.vehicle as vehicle

from sim.subsystems import Simulation
from sim.vision import IBuoy, IPipe, IBarbedWire, IBin, ITarget, ISafe

class PerfectSlam(slam.ISlam):
    def __init__(self, config, deps):
        slam.ISlam.__init__(self, config.get('name', 'Slam'),
                            core.Subsystem.getSubsystemOfType(
                core.QueuedEventHub, deps))

        # Grab the vehicle and the controller
        self._vehicle = core.Subsystem.getSubsystemOfType(vehicle.IVehicle,
                                                          deps,
                                                          nonNone = True)
        self._controller = core.Subsystem.getSubsystemOfType(
            control.IController, deps, nonNone = True)

        # Create the dictionary that stores the values
        self._objects = {}

        # Grab the simulation object
        sim = core.Subsystem.getSubsystemOfType(Simulation, deps, 
                                                    nonNone = True)

        self._getObjectInstances(sim, "buoy", IBuoy)
        self._getObjectInstances(sim, "pipe", IPipe)
        self._getObjectInstances(sim, "bwire", IBarbedWire)
        self._getObjectInstances(sim, "bin", IBin)
        self._getObjectInstances(sim, "target", ITarget)
        self._getObjectInstances(sim, "safe", ISafe)

    def _getObjectInstances(self, sim, name, interface):
        robot = self._vehicle.getDevice('SimulationDevice').robot
        robotPos = robot._main_part._node.position

        objects = sim.scene.getObjectsByInterface(interface)
        if objects is None or len(objects) == 0:
            return
        if len(objects) == 1:
            position = objects[0].position
            self._objects[name] = math.Vector2(position.x - robotPos.x,
                                               position.y - robotPos.y)
        else:
            for i, obj in enumerate(objects):
                self._objects[name + str(i+1)] = \
                    math.Vector2(obj.position.x - robotPos.x,
                                 obj.position.y - robotPos.y)

    def getObjectPosition(self, name):
        return self._objects[name]

    def getRelativePosition(self, name):
        return self._vehicle.getPosition() - self._objects[name]

    def hasObject(self, name):
        return self._objects.has_key(name)

    def backgrounded(self):
        return True

    def unbackground(self, join = False):
        pass

    def background(self, interval = -1):
        pass

core.SubsystemMaker.registerSubsystem('PerfectSlam', PerfectSlam)
